packages/grafana-prometheus/src/datasource.ts

```diff
@@ -1041,13 +1045,49 @@ export function extractRuleMappingFromGroups(groups: RawRecordingRules[]): RuleQ
   );
 }
 
-// NOTE: these two functions are very similar to the escapeLabelValueIn* functions
-// in language_utils.ts, but they are not exactly the same algorithm, and we found
-// no way to reuse one in the another or vice versa.
+// NOTE: `prometheusRegularEscape` and `prometheusSpecialRegexEscape` are similar
+// to the escapeLabelValueIn* functions in language_utils.ts, but they are not exactly
+// the same algorithm, and we found no way to reuse one in the another or vice versa.
 export function prometheusRegularEscape<T>(value: T) {
-  return typeof value === 'string' ? value.replace(/\\/g, '\\\\').replace(/'/g, "\\\\'") : value;
+  if (typeof value !== 'string') {
+    return value;
+  }
+
+  return (
+    value
+      // escape backslashes
+      .replace(/\\/g, '\\\\')
+      // escape single quotes with double backslashes
+      .replace(/'/g, "\\\\'")
```

Comment:
Single quotes don't need to be escaped here, but double quotes do:

```
// escape double quotes with single backslash
.replace(/"/g, '\\"')
```

The proposed code would escape non-regex promql queries like this:

```
singlequote_escape_test{label="looking\\'glass"}
doublequote_escape_test{label="looking"glass"}
```

The doublequote_escape_test query produces a syntax error from prometheus:

```
bad_data: invalid parameter "query": 1:40: parse error: unexpected identifier "glass" in label matching, expected "," or "}"
```

The correct queries should be:

```
singlequote_escape_test{label="looking'glass"} 1
doublequote_escape_test{label="looking\"glass"} 1
```